Using auth docs (#1359)
wasp-lang/wasp: web/docs/auth/overview.md

title: Using Auth
---

import { AuthMethodsGrid } from "@site/src/components/AuthMethodsGrid";
import { Required } from "@site/src/components/Required";

Auth is an important piece of each more advanced application. Wasp luckily provides authentication and authorization support out of the box 🙃

Enabling it for your app is optional and can be done by configuring the `auth` field of the `app` declaration.  

<Tabs groupId="js-ts">
<TabItem value="js" label="JavaScript">

> ```wasp

Author: I usually put `wasp.main`, I must have missed this one 👍

We will also have access to the `user` object in our frontend code, so we can show different UI to logged-in and logged-out users. For example, we can show the user's name in the header alongside a **logout button** or a login button if the user is not logged in.

## Protecting a page with `authRequired`

When you declare a `page` you can provide it with `authRequired` property.

If you set it to `true`, only authenticated users will be able to access this page. Unauthenticated users will be redirected to a route defined by `onAuthFailedRedirectTo` property within `app.auth`. 

<Tabs groupId="js-ts">
<TabItem value="js" label="JavaScript">

```wasp
page MainPage {
  component: import Main from "@client/pages/Main.jsx",
  authRequired: true
}
```
</TabItem>
<TabItem value="ts" label="TypeScript">

```wasp
page MainPage {
  component: import Main from "@client/pages/Main.tsx",
  authRequired: true
}
```
</TabItem>
</Tabs>

:::caution Requires auth method

> `authRequired` can only be used if some auth method is enabled for the app.

Reviewer: A strange sentence (probably due to the passive voice), I suggest:

```suggestion
You can only use `authRequired` if your app uses one of the [available auth methods](#available-auth-methods).
```

:::

> If `authRequired` is set to `true`, the React component of the page (specified by `component` property) will be provided `user` object as a prop. Read more about the `user` object in the [Accessing the logged-in user](#accessing-the-logged-in-user) section.

Reviewer: ```suggestion
If `authRequired` is set to `true`, the page's React component (specified by the `component` property) receives the `user` object as a prop. Read more about the `user` object in the [Accessing the logged-in user section](#accessing-the-logged-in-user).
```

## Logout action

We provide an action for logging out the user. Here's how you can use it:

<Tabs groupId="js-ts">
<TabItem value="js" label="JavaScript">

```jsx title="client/components/LogoutButton.jsx"
import logout from '@wasp/auth/logout'

const LogoutButton = () => {
  return (
    <button onClick={logout}>Logout</button>
  )
}
```
</TabItem>
<TabItem value="ts" label="TypeScript">

```tsx title="client/components/LogoutButton.tsx"
import logout from '@wasp/auth/logout'

const LogoutButton = () => {
  return (
    <button onClick={logout}>Logout</button>
  )
}
```
</TabItem>
</Tabs>

## Accessing the logged-in user

You can get access to the `user` object both in the backend and on the frontend.

### On the client

#### Using the `user` prop

> If `authRequired` is set to `true` for a `page` declaration, the React component of the page will be provided `user` object as a prop:

Reviewer: ```suggestion
If the page's declarations sets `authRequired` to `true`, the page's React component receives the `user` object as a prop:
```

<Tabs groupId="js-ts">
<TabItem value="js" label="JavaScript">

```wasp title="main.wasp"
// ...

page AccountPage {
  component: import Account from "@client/pages/Account.jsx",
  authRequired: true
}
```

```jsx title="client/pages/Account.jsx"
import Button from './Button';
import logout from '@wasp/auth/logout';

const AccountPage = ({ user }) => {
  return (
    <div>
      <Button onClick={logout}>Logout</Button>
      {JSON.stringify(user, null, 2)}
    </div>
  );
};

export default AccountPage;
```
</TabItem>
<TabItem value="ts" label="TypeScript">

```wasp title="main.wasp"
// ...

page AccountPage {
  component: import Account from "@client/pages/Account.tsx",
  authRequired: true
}
```

```tsx title="client/pages/Account.tsx"
import type { User } from '@wasp/entities';
import Button from './Button';
import logout from '@wasp/auth/logout';

const AccountPage = ({ user }: { user: User }) => {
  return (
    <div>
      <Button onClick={logout}>Logout</Button>
      {JSON.stringify(user, null, 2)}
    </div>
  );
};

export default AccountPage;
```
</TabItem>
</Tabs>

> #### Using the `useAuth` hook

Reviewer: Can we explain when they should use the prop as opposed to the hook (and vice versa)?

<TabItem value="js" label="JavaScript">

```jsx title="src/client/pages/MainPage.jsx"
import useAuth from '@wasp/auth/useAuth'
import { Link } from 'react-router-dom'
import logout from '@wasp/auth/logout'
import Todo from '../Todo'

export function Main() {
  const { data: user } = useAuth()

  if (!user) {
    return (
      <span>
        Please <Link to='/login'>login</Link> or <Link to='/signup'>sign up</Link>.
      </span>
    )
  } else {
    return (
      <>
        <button onClick={logout}>Logout</button>
        <Todo />
      < />
    )
  }
}
```
</TabItem>
<TabItem value="ts" label="TypeScript">

```tsx title="src/client/pages/MainPage.tsx"
import useAuth from '@wasp/auth/useAuth'
import { Link } from 'react-router-dom'
import logout from '@wasp/auth/logout'
import Todo from '../Todo'

export function Main() {
  const { data: user } = useAuth()

  if (!user) {
    return (
      <span>
        Please <Link to='/login'>login</Link> or <Link to='/signup'>sign up</Link>.
      </span>
    )
  } else {
    return (
      <>
        <button onClick={logout}>Logout</button>
        <Todo />
      < />
    )
  }
}
```
</TabItem>
</Tabs>

### On the server

> #### Using the `context.user`

Reviewer: I think this title should either be "Using `context.user` or "Using the `context.user` object".

> When authentication is enabled, all [queries and actions](/docs/database/operations) will have access to the `user` through the `context` argument. `context.user` will contain all the fields from the user entity except for the password.

Reviewer: Personal suggestion: it's best to avoid future tenses when possible.
```suggestion
When authentication is enabled, all [queries and actions](/docs/database/operations) have access to the `user` object through the `context` argument. `context.user` contains all User entity's fields, except for the password.
```

<Tabs groupId="js-ts">
<TabItem value="js" label="JavaScript">

```js title="src/server/actions.js"
import HttpError from '@wasp/core/HttpError.js'

export const createTask = async (task, context) => {
  if (!context.user) {
    throw new HttpError(403)
  }

  const Task = context.entities.Task
  return Task.create({
    data: {
      description: task.description,
      user: {
        connect: { id: context.user.id }
      }
    }
  })
}
```
</TabItem>
<TabItem value="ts" label="TypeScript">

```ts title="src/server/actions.ts"
import type { Task } from "@wasp/entities"
import type { CreateTask } from "@wasp/actions/types"
import HttpError from '@wasp/core/HttpError.js'

type CreateTaskPayload = Pick<Task, "description">

export const createTask: CreateTask<CreateTaskPayload, Task> = async (args, context) => {
  if (!context.user) {
    throw new HttpError(403)
  }

  const Task = context.entities.Task
  return Task.create({
    data: {
      description: args.description,
      user: {
        connect: { id: context.user.id }
      }
    }
  })
}
```
</TabItem>
</Tabs>

> To implement access control, each operation is responsible for checking `context.user` and acting accordingly - e.g. if `context.user` is `undefined` and the operation is private then user should be denied access to it.

Reviewer: ```suggestion
Make sure to take care of Access-control. Each operation is responsible for checking `context.user` and acting accordingly. For example, if `context.user` is `undefined` inside a private operation, the user's access should be denied.
```

> When using WebSockets, the `user` object is also available on the `socket.data`. Read more in the [WebSockets](/docs/advanced/web-sockets#websocketfn-function) section.

Reviewer: ```suggestion
When using WebSockets, the `user` object is also available on the `socket.data` object. Read more in the [WebSockets section](/docs/advanced/web-sockets#websocketfn-function).
```

## User entity

### Password hashing

> You don't need to worry about hashing the password yourself. Even when you are using Prisma's client directly and calling `create()` with a plain-text password, Wasp's middleware takes care of hashing it before storing it in the database.
> 
> For example, if you need to update a user's password, you can do it safely via Prisma client, e.g. within an action:

Reviewer: ```suggestion
You don't need to worry about hashing the password yourself. Even when directly using the Prisma client and calling `create()` with a plain-text password, Wasp's middleware makes sure to hash the password before storing it in the database.
For example, if you need to update a user's password, you can safely use the Prisma client to do so, e.g., inside an Action:
```

<Tabs groupId="js-ts">
<TabItem value="js" label="JavaScript">

```js title="src/server/actions.js"
export const updatePassword = async (args, context) => {
  return context.entities.User.update({
    where: { id: args.userId },
    data: {
      password: 'New pwd which will be hashed automatically!'
    }
  })
}
```
</TabItem>
<TabItem value="ts" label="TypeScript">

```ts title="src/server/actions.ts"
import type { UpdatePassword } from "@wasp/actions/types"
import type { User } from "@wasp/entities"

type UpdatePasswordPayload = {
  userId: User["id"]
}

export const updatePassword: UpdatePassword<UpdatePasswordPayload, User> = async (args, context) => {
  return context.entities.User.update({
    where: { id: args.userId },
    data: {
      password: 'New pwd which will be hashed automatically!'
    }
  })
}
```
</TabItem>
</Tabs>

### Default validations

> We provide basic validations out of the box, which you can customize as shown below. 

Reviewer: Since we're using "we" as an entity going through the guide (i.e., us + the reader), we should refrain from using it when talking about Wasp.

```suggestion
Wasp includes several basic validation mechanisms. If you need something extra, the [next section](#customizing-validations) shows how to customize them.
```

Default validations depend on the auth method you use.

#### Username & password

> If you use the [Username & password](docs/auth/username-and-pass) authentication, the default validations are:

Reviewer: Dead link, missing leading slash.

The `the` before "Username & password" should be removed.

- `username`: non-empty

> - `password`: non-empty, at least 8 characters, and contains a number

Reviewer: I suggest formulating these as sentences, for example: The `username` must not be empty.

Note that `username`s are stored in a **case-sensitive** manner.

#### Email

> If you use the [Email](docs/auth/email) authentication, the default validations are:

Reviewer: ```suggestion
If you use [Email](docs/auth/email) authentication, the default validations are:
```

Same comment about the bullet points (i.e., I think we should turn them into sentences).

Also, dead link (leading slash).

  },
  _waspSkipDefaultValidations: false, // can be omitted if false (default), or explicitly set to true
  _waspCustomValidations: [
    {
      validates: 'password',
      message: 'password must contain an uppercase letter',
      validator: password => /[A-Z]/.test(password)
    },
  ]
})
```
</TabItem>
<TabItem value="ts" label="TypeScript">

```ts
const newUser = context.entities.User.create({
  data: {
    username: args.username,
    password: args.password // password hashed automatically by Wasp! 🐝
  },
  _waspSkipDefaultValidations: false, // can be omitted if false (default), or explicitly set to true
  _waspCustomValidations: [
    {
      validates: 'password',
      message: 'password must contain an uppercase letter',
      validator: password => /[A-Z]/.test(password)
    },
  ]
})
```
</TabItem>
</Tabs>

:::info
Validations always run on `create()`, but for `update()` they only run when the field mentioned in `validates` is present.

The validation process stops on the first `validator` to return false. If enabled, default validations run first and then custom validations.
:::

### Validation Error Handling

> When creating, updating, or deleting entities, you may wish to handle validation errors. We have exposed a class called `AuthError` for this purpose.

Reviewer: ```suggestion
When creating, updating, or deleting entities, you may wish to handle validation errors. Wasp exposes a class called `AuthError` for this purpose.
```

  provider    String
  providerId  String
  user        User      @relation(fields: [userId], references: [id], onDelete: Cascade)
  userId      Int
  createdAt   DateTime  @default(now())
  @@unique([provider, providerId, userId])
psl=}
```
</TabItem>
<TabItem value="ts" label="TypeScript">

```wasp {4,14}
//...
  auth: {
    userEntity: User,
    externalAuthEntity: SocialLogin,
//...

entity User {=psl
    id                        Int           @id @default(autoincrement())
    //...
    externalAuthAssociations  SocialLogin[]
psl=}

entity SocialLogin {=psl
  id          Int       @id @default(autoincrement())
  provider    String
  providerId  String
  user        User      @relation(fields: [userId], references: [id], onDelete: Cascade)
  userId      Int
  createdAt   DateTime  @default(now())
  @@unique([provider, providerId, userId])
psl=}
```
</TabItem>
</Tabs>

:::note
The same `externalAuthEntity` can be used across different social login providers (e.g., both GitHub and Google can use the same entity).
:::

See [Google docs](/docs/auth/google) and [GitHub docs](/docs/auth/github) for more details.

#### `methods: dict` <Required />

> A dictionary of auth methods that are enabled for the app.

Reviewer: ```suggestion
A dictionary of auth methods enabled for the app.
```

<AuthMethodsGrid />

#### `onAuthFailedRedirectTo: String` <Required />

> The path where an unauthenticated user will be redirected if they try to access a private page (which is declared by setting `authRequired: true` for a specific page).

Reviewer: ```suggestion
The route to which Wasp should redirect unauthenticated user when they try to access a private page (i.e., a page that has `authRequired: true`).
```

Check out these [essentials docs on auth](/docs/essentials/auth#adding-auth-to-the-project) to see an example of usage.

#### `onAuthSucceededRedirectTo: String`

> The path where a successfully authenticated user will be sent upon successful login/signup.

Reviewer: ```suggestion
The route to which Wasp will send a successfully authenticated after a successful login/signup.
```

The default value is `"/"`.

:::note

> Automatic redirect on successful login only works when using the Wasp provided [`Signup` and `Login` forms](#high-level-api)

Reviewer: ```suggestion
Automatic redirect on successful login only works when using the Wasp-provided [`Signup` and `Login` forms](#high-level-api)
```

> Automatic redirect on successful login only works when using the Wasp provided [`Signup` and `Login` forms](#high-level-api)

Reviewer: Dead link, but don't know what it's supposed to be.
